ts/src/graphql/query/edge_connection.ts

```diff
@@ -74,23 +76,34 @@ export class GraphQLEdgeConnection<
       if (this.args.before && !this.args.before) {
         throw new Error("cannot process before without last");
       }
+      const argFirst = this.args.first;
+      const argLast = this.args.last;
+      const argCursor = this.args.cursor;
       if (this.args.first) {
-        this.query = this.query.first(this.args.first, this.args.after);
+        this.query = Promise.resolve(this.query).then((query) =>
+          query.first(argFirst, argLast),
+        );
       }
       if (this.args.last) {
-        this.query = this.query.last(this.args.last, this.args.cursor);
+        this.query = Promise.resolve(this.query).then((query) =>
+          query.last(argLast, argCursor),
+        );
       }
       // TODO custom args
       // how to proceed
     }
   }
 
   first(limit: number, cursor?: string) {
-    this.query = this.query.first(limit, cursor);
+    this.query = Promise.resolve(this.query).then((query) =>
+      query.first(limit, cursor),
+    );
   }
 
   last(limit: number, cursor?: string) {
-    this.query = this.query.last(limit, cursor);
+    this.query = Promise.resolve(this.query).then((query) =>
+      query.last(limit, cursor),
+    );
```

Comment:
hmm. what's the usecase?

is this idiomatic TS?

do we want to use Promise.resolve here? doesn't seem ideal

Reply:
> what's the usecase?

I have a custom edge query where the user can pass in a location string and I fetch the lat/lon before running the query, so I need that method to support returning a Promise.

> is this idiomatic TS?

Not entirely sure what you mean, but it's normal JS here.

> do we want to use Promise.resolve here? doesn't seem ideal

I couldn't think of a better way to do it since `EdgeQuery.first` and `EdgeQuery.last` are functions that mutated the object. I don't think it's _unideal_ to use `Promise.resolve` here. If you call it more than once it's a no-op, and if you call it on the non-promise it's just an immediate return.

Reply:
basically we have a non async function here which resolves a Promise and doesn't play well with the event loop.

is this worth a breaking change or a different API for this case?